pkg/client/oci/oci.go

```diff
@@ -0,0 +1,91 @@
+package oci
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"github.com/jetstack/version-checker/pkg/api"
+	"github.com/jetstack/version-checker/pkg/version/semver"
+)
+
+type Client struct{}
+
+func New() (*Client, error) {
+	return &Client{}, nil
+}
+
+func (c *Client) Name() string {
+	return "oci"
+}
+
+func (c *Client) Tags(ctx context.Context, host, repo, image string) ([]api.ImageTag, error) {
+	src := fmt.Sprintf("%s/%s/%s", host, repo, image)
+	rpo, err := name.NewRepository(src)
+
+	if err != nil {
+		return []api.ImageTag{}, err
+	}
+
+	bareTags, err := remote.List(rpo, remote.WithContext(ctx))
+
+	if err != nil {
+		return []api.ImageTag{}, err
+	}
+
+	sort.SliceStable(bareTags, func(i, j int) bool {
+		return semver.Parse(bareTags[i]).LessThan(semver.Parse(bareTags[j]))
+	})
+
+	var tags []api.ImageTag
+	for i, t := range bareTags {
+		tag := api.ImageTag{Tag: t}
+
+		// Only fetch metadata for the highest version
+		// Could be very slow otherwise if there are a lot of tags, and
+		// we only really care about the highest version.
```

Comment:
Forgive my ignorance of version-checker here, but is this definitely true? Isn't it possible that there will be some custom sorting of the `[]api.ImageTag` this function returns that will require full details? For instance, finding the latest image by `Timestamp`?

The cost of fetching the config for every tag may just be the cost of doing business here?

I suppose if we want to make it more efficient, we could cache details by digest? That way we'd only ever have to fetch the config for each digest once?

Reply:
Hm, looking at it harder, I missed that there's a `use-sha` option, where it checks against the latest sha by date. I kinda dismissed that as something that people might want to do, because the latest by date might often be a hotfix against an older version. But I guess there is a use-case for if people are using `latest` or something - although then I'd expect them to either flush their pods regularly, or have some CI around it.

> The cost of fetching the config for every tag may just be the cost of doing business here?

I tested against the fluentbit repo, I think it hadn't finished after an hour. I think if we were talking a few minutes I'd accept the cost, but potentially for repos with large numbers of tags and api rate limits, we could be talking days.

Reply:
Hm, in that case, maybe we just don't fetch the config at all?

Reply:
The prom metrics output wants to give you the sha of the latest version.

I think the best thing to do here is to fetch the metadata for the latest semver, and for any currently deployed tags. Still a gap for empty tags, but that seems like an edge case.

Reply:
Maybe we load the highest semver, and any non-semver tags - that'd be easier to implement, hopefully should be a reasonably small set, and I think would cater for all use cases.

Reply:
Had a chat with @ribbybibby and think it's better to skip the metadata for now. There's a few different ways in which we consider a version to be "current", and it's not trivial to pull that logic in here.